interpreter.js

```diff
@@ -2073,6 +2073,7 @@ Interpreter.prototype.initRegExp = function(globalObject) {
         // match has additional properties.
         thisInterpreter.setProperty(result, 'index', match.index);
         thisInterpreter.setProperty(result, 'input', match.input);
+        thisInterpreter.setProperty(result, 'groups', thisInterpreter.nativeToPseudo(match.groups));
```

Comment:
There is no such `groups` property [mentioned in the ECMAScript 5.1 spec for `RegExp.prototype.exec`](https://262.ecma-international.org/5.1/#sec-15.10.6.2).

Reply:
It's not even in ES8: https://ecma-international.org/wp-content/uploads/ECMA-262_8th_edition_june_2017.pdf
It first shows up in ES9: https://ecma-international.org/wp-content/uploads/ECMA-262_9th_edition_june_2018.pdf

In JS-Iterpreter both `match` and `exec` are currently returning arrays with `groups` properties.  But both are being created by the native function in the user's browser.

```
var m = '123'.match(/(\d)/);
alert(Object.keys(m));
var e = /(\d)/.exec('123');
alert(Object.keys(e));
```

Since we are an ES5 interpreter, my thinking is that we should delete these properties.  Thoughts?

Reply:
Deleting `groups` would work but it's not robust against future enhancements of `RegExp.prototype.exec`.  I can't easily check TC39 at the moment but it would be surprising if there were no such proposals.

Fundamentally `matchToPseudo` is unsafe because it passes something which is not guaranteed to be a simple array to `arrayNativeToPseudo`, which trusts that its argument's properties are all either primitives or already PsedoObjects.

Either `arrayNativeToPseudo` should check this invariant at runtime (slow, but safest!) or you should audit all calls to make sure that none can violate it.

Reply:
I've commited c3f7c00641c66ce94028f7a71c9a36ea18013e15 which filters the properties of `match` and `exec` to only have ES5 properties.  It also normalizes `match` and `exec`, which previously were each doing their own thing, differently.  I think this resolves the primary issue.

However, this makes me look critically at `arrayNativeToPseudo` and `arrayPseudoToNative`.  I don't think these functions should exist.  They duplicate the functionality of `nativeToPseudo` and `pseudoToNative` respectively.  The array methods are unsafe since they aren't recursive, but on the other hand they are faster.  However, neither are used in high-traffic areas (Object.keys, Object.getOwnPropertyNames, Function.apply, String.split, String.match, RegExp.exec, JSON.parse), so I don't think there's a good argument for them.

Reply:
I've commited cba01c8468297f0068a753e9abfe37364df6ab8c which overhauls `nativeToPseudo` and `pseudoToNative`.  Lots of improvements, they are now mutually consistent.

`arrayNativeToPseudo` is completely gone.  However `arrayPseudoToNative` remains due to one usage in Function.apply which requires non-recursive conversion.

I think these changes resolves the issues raised in this PR without introducing any new issues.

Reply:
Thank you for your response and discussion.
And thanks for the detailed explanation.

Is it acceptable to close this PR?

Reply:
The third commit a6076c6da554e00764efa81fab7ecc118e04f9d6 cleans up Function's `.apply`.  Also completed an experiment to see if optional use of Map would help speed up cycle detection (it did not).  That clears the last questions regarding this topic.

I think we're are good stable state now.  Closing.  Thanks for raising this issue!